### PR TITLE
feat: support add custom tempdir for modern.js App

### PR DESCRIPTION
**Pull request by zllkjc**
## Summary

Modern.js will create tempDir for the real webpack entry before build. If there is multiple config in one app, they can not build together. Now we support a new params for `@modern-js/app-tools`:

```ts
import path from 'path';
import appTools, { defineConfig } from '@modern-js/app-tools';

export default defineConfig({
  output: {
    tempDir: path.join('node_modules', '.temp-dir'),
  },
  plugins: [appTools({})],
});

```

## Checklist

- [x] I have added changeset via `pnpm run change`.
- [x] I have updated the documentation.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: packages/solutions/app-tools/src/index.ts [hunk 1]
```diff
@@ -154,6 +154,19 @@ export default (
     i18n.changeLanguage({ locale });
 
     return {
+      async afterInit() {
```
Comment (chenjiahan) [inline packages/solutions/app-tools/src/index.ts:157]: How about running the following codes in the body of `setup` function:

image

Comment (chenjiahan) [reply packages/solutions/app-tools/src/index.ts:157]: Oh It doesn't seem to work, the AppContext has not been initialized when the setup is executed,

File: packages/solutions/app-tools/src/types/config/output.ts [hunk 2]
```diff
@@ -14,6 +14,7 @@ export interface SharedOutputConfig extends BuilderSharedOutputConfig {
   ssg?: SSGConfig;
   splitRouteChunks?: boolean;
   disableNodePolyfill?: boolean;
+  tempDir?: string;
```
Comment (chenjiahan) [inline packages/solutions/app-tools/src/types/config/output.ts:17]: The `output.tempDir` config should be documented.

File: packages/solutions/app-tools/src/index.ts (hunk 1 above)
Comment (chenjiahan) [reply packages/solutions/app-tools/src/index.ts:157]: How about renaming `afterInit` to `beforeConfig`, because there is no `init` hook

Comment (zllkjc) [reply packages/solutions/app-tools/src/index.ts:157]: Okey, that's better~

Comment (chenjiahan) [reply packages/solutions/app-tools/src/index.ts:157]: The `beforeConfig` should also be documented 😄

Comment (zllkjc) [reply packages/solutions/app-tools/src/index.ts]: done~

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**